Add command to print routes

francisbrito opened the pull request:
Hello,

I found myself needing a way to see a list (or tree) breakdown of all the routes my project had and thought it might be useful to have a command in `fastify-cli` that does just that.

This is essentially a CLI wrapper for `fastify.printRoutes`.

It behaves similarly to `fastify start` command except it relies on `fastify.ready()` rather than `fastify.listen()`, as such, there are a few function that are identical in both commands and could probably be abstracted and reused. If it seems reasonable, I could do it in the same PR. 

The reason I didn't write these abstractions is I preferred not to introduce too many changes to the project structure in order not to make this PR too invasive.

mcollina commented on print-routes.js:
```diff
@@ -0,0 +1,116 @@
+
+if (require.main === module) {
+  cli(process.argv.slice(2))
+}
```
Can you refactor the logic in this file and the start command to reduce duplication?

francisbrito replied:
Sure thing!

I have created `util.js` just for this. Please have a look and let me know what you think.

mcollina commented on print-routes.js:
```diff
@@ -0,0 +1,116 @@
+}
+
+function cli (args) {
+  printRoutes(args)
```
I think we should call fastify.close() after having printed the routes. It’s important so the application would stop gracefully.

francisbrito replied:
Good catch.

I added a call to `fastify.close` in my latest commit.

francisbrito commented:
In addition to applying the improvements highlighted in the review, I added the an entry for the `print-routes` command to the `help` command output.

Also fixed tests failing in Node 6. (I hadn't realized Travis CI check was failing).

xtx1130 commented on start.js:
```diff
@@ -4,34 +4,33 @@
   }
 }
 
 function showHelp () {
```
I think it is unnecessary to make `showHelpForCommand` inside `showHelp`. You can use `exit()` inside `showHelpForCommand`, and delete `showHelp` there.

xtx1130 commented on util.js:
```diff
@@ -0,0 +1,72 @@
+
+    return { module, pkg }
+  } catch (e) {
+    console.log('Warn: unable to load fastify module')
```
```suggestion
    exit('unable to load fastify module')
```

xtx1130 commented on util.js:
```diff
@@ -23,8 +23,8 @@ function requireFastifyForModule (modulePath) {
-    console.log('Warn: unable to load fastify module')
-    throw e
+    console.log('unable to load fastify module')
+    exit()
```
I think you can use `exit(message)` instead of `console.log(message)`

xtx1130 commented on util.js:
```diff
@@ -63,8 +63,9 @@ function showHelpForCommand (commandName) {
 
   try {
     console.log(fs.readFileSync(helpFilePath, 'utf8'))
+    exit()
```
And there

francisbrito replied:
I think changing this line to be `exit(fs.readFileSync(helpFilePath, 'utf8'))` would cause help command to return a non-zero exit code. Is this OK?

I might be misinterpreting your change request, so please do let me know if so is the case.

xtx1130 replied:
@francisbrito I am sorry for my mistake, I wanna add comment on [Line67](https://github.com/fastify/fastify-cli/pull/193/files#diff-63f3aa32c5cc5d42a250b081693e9dc2R67). I think is unnecessary to tell users two messages about the error

xtx1130 commented on util.js:
```diff
@@ -0,0 +1,72 @@
+    console.log(fs.readFileSync(helpFilePath, 'utf8'))
+    exit()
+  } catch (e) {
+    console.log(`unable to get help for command "${commandName}"`)
```
Here🙋‍♂️。
```suggestion
    exit(new Error(`unable to get help for command "${commandName}"`))
```

francisbrito replied:
My reasoning for adding the extra `console.log` comment was adding some context on what is the underlying cause of the error. But thinking it better, if a command errors when the user passes the `--help` flag the error would be out of the hands of their hands.

mcollina commented on print-routes.js:
```diff
@@ -0,0 +1,86 @@
+  } catch (e) {
+    module.exports.stop(e)
+  }
+}
```
I think this can be abstracted away and shared with start.

francisbrito replied:
Like this?

```js
// util.js
function tryRequireFastifyForModuleOrExit(modulePath) {
  try {
    return requireFastifyForModule(modulePath);
  } catch (e) {
    exit(e);
  }
}
```

mcollina replied:
I was wrong. It's fine as it is.

Eomm commented on test/start.test.js:
```diff
@@ -209,17 +209,21 @@ test('should start the server with an async/await plugin', t => {
 
-test('should exit without error on help', t => {
-  t.plan(1)
+t.only('should exit without error on help', t => {
```
```suggestion
test('should exit without error on help', t => {
```
